CONTRIBUTING.md

```diff
@@ -188,7 +188,57 @@ with one another are rolled up.
 Speaking of tests, Rust has a comprehensive test suite. More information about
 it can be found [here][rctd].
 
-### External Dependencies
+### External Dependencies (subtree)
+
+As a developer to this repository, you don't have to treat the following external projects
+differently from other crates that are directly in this repo:
+
+* none so far, see https://github.com/rust-lang/rust/issues/70651 for more info
+
+They are just regular files and directories. This is in contrast to `submodule` dependencies
+(see below for those). Only tool authors will actually use any operations here.
+
+#### Synchronizing a subtree
+
+There are two synchronization directions: `subtree push` and `subtree pull`.
+
+`git subtree push -P src/tools/clippy https://github.com/rust-lang/rust-clippy.git`
+
+takes all the changes that
+happened to the copy in this repo and creates commits on the remote repo that match the local
```

Comment:
So push (and also pull) both create one commit "on the other side" for each commit "on this side"? There is no squashing in either direction?

Reply:
Yes, no squashing at all. In the pull direction we apparently even preserve commit hashes.

Reply:
AFAICT, each commit is copied to the other side verbatim (even the hash is the same, at least for pull, not sure about push), then a single "subtree merge commit" is added which places the changes at the right prefix (when pulling) or removes the prefix (when pushing).

Reply:
Sounds good. That might be worth mentioning in the docs as well.

Reply:
Not sure what is missing from the docs. I've stated that push commits get modified and pull commits get copied. Do you want me to rehash this in detail? Seems to me like this would be useless noise to the person looking for usage instructions. I'd just be moving `git subtree` documentation into our documentation in a trimmed and incomplete way.

Reply:
Linking to well-written existing `git subtree` documentation is certainly also an option. :)

I just feel like one should be able to learn the results of the experiments that @eddyb described above by reading this document and following links.

Reply:
I understand that we need this information to make a decision, but I don't see how it helps us to have this documentation in the contribution guide.

Reply:
I think it's good to understand the tool one is using, and this is a key part to understanding how `git subtree` works. It is the kind of knowledge you need to debug things when things get weird (and knowing git, they *will* get weird sometimes).

Reply:
I added more docs